File: mixins/powers_meta.rb
require 'byebug'


def power(&block)
  Module.new do
    def self.included(base)
      def max_speed(speed)
        @max_speed = speed
      end

      instance_eval(&block)
    end
  end
end

Flyable = power do
  max_speed 10
end

Leapable = power do
  max_speed 6
end

Runnable = power do
  max_speed 4
end

Walkable = power do
  max_speed 1
end

class BaseHero
  include Walkable
end

class SortaFastHero < BaseHero
  include Runnable
end

class FasterHero < BaseHero
  include Leapable
end

class CuriouslySlowHero < BaseHero
  include Flyable
  include Runnable
end

class FastestHero < BaseHero
  include Flyable
end

# An imaginary game board that doesn't do anything.
class Board
  # Move a piece on the board.
  #
  # ``piece`` should be movable
  # ``actions`` is the number of consecutive actions taken to move
  #
  # Returns the total space that ``piece`` moved on the board.
  def move(piece, actions_spent)
    # Fictitious piece-moving machinery here
    actions_spent * piece.max_speed
  end
end

board = Board.new

byebug

puts

heroes = [
  ['Hero', 'Total spaces moved'],
  ['a sorta fast hero', board.move(SortaFastHero.new, 2)],
  ['a faster hero', board.move(FasterHero.new, 2)],
  ['a curiously slow hero', board.move(CuriouslySlowHero.new, 2)],
  ['the fastest hero', board.move(FastestHero.new, 2)]
]

heroes.each do |description, movement|
  printf "\t%-22s %25s\n", description, movement
end

File: mixins/powers_meta.py
class FlyMixin:
    @property
    def max_speed(self):
        return 10


class GiantLeapMixin:
    @property
    def max_speed(self):
        return 6


class RunMixin:
    @property
    def max_speed(self):
        return 4


class BaseHero:
    @property
    def max_speed(self):
        return 1


class SortaFastHero(RunMixin, BaseHero):
    """This hero can run, which is better than nothing."""
    pass


class FasterHero(GiantLeapMixin, RunMixin, BaseHero):
    """This hero can leap pretty quickly."""
    pass


class CuriouslySlowHero(BaseHero, FlyMixin):
    """This hero is confused about her powers."""
    pass


class FastestHero(FlyMixin, BaseHero):
    """The fastest hero can fly, of course."""
    pass


class Board:
    """An imaginary game board that doesn't do anything."""
    def move(self, piece, actions_spent):
        """Move a piece on the board.

        ``piece`` should be movable
        ``actions`` is the number of consecutive actions taken to move

        Returns the total space that ``piece`` moved on the board.
        """
        # Fictitious piece-moving machinery here
        return actions_spent * piece.max_speed


def main():
    board = Board()
    heroes = (
        ('Hero', 'Total spaces moved'),
        ('a sorta fast hero', board.move(SortaFastHero(), 2)),
        ('a faster hero', board.move(FasterHero(), 2)),
        ('a curiously slow hero', board.move(CuriouslySlowHero(), 2)),
        ('the fastest hero', board.move(FastestHero(), 2))
    )
    print('\n')
    for description, movement in heroes:
        print("\t{:<22} {:>25}".format(description, movement))


if __name__ == '__main__':
    main()
